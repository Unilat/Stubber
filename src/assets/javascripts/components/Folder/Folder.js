import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { createSelector } from 'reselect';
import classNames from 'classnames';
import { folderActions } from '../../reducers/folders';
import { addStub, disableStub } from '../../reducers/stubs';
import Stub from '../Stub';
import Tooltip from '../Tooltip';
import RemoveButton from '../RemoveButton';
import Modal from '../Modal';
import StubForm from '../StubForm';

import './Folder.scss';

export class Folder extends Component {

    static propTypes = {
        folder: PropTypes.object.isRequired,
        editing: PropTypes.bool.isRequired,
        actions: PropTypes.object.isRequired,
        disabled: PropTypes.bool.isRequired,
        id: PropTypes.number.isRequired
    };

    constructor(props) {
        super(props);

        this.state = {
            addStubOpen: false,
            editing: this.props.editing
        };

        this.addStub = this.addStub.bind(this);
        this.editName = this.editName.bind(this);
        this.nameBlur = this.nameBlur.bind(this);
        this.nameInput = this.nameInput.bind(this);
        this.remove = this.remove.bind(this);
        this.closeAddStub = this.closeAddStub.bind(this);
        this.toggle = this.toggle.bind(this);
        this.inputClick = this.inputClick.bind(this);
        this.disable = this.disable.bind(this);
        this.mouseDown = this.mouseDown.bind(this);
        this.mouseMove = this.mouseMove.bind(this);
        this.mouseUp = this.mouseUp.bind(this);

        this.app = document.getElementsByClassName('stubber')[0];
    }

    componentDidMount() {
        if (this.state.editing) {
            this.editName();
        }

        this.updateDropdown();
    }

    componentDidUpdate() {
        this.updateDropdown();
    }

    shouldComponentUpdate(nextProps, nextState) {
        // only update when folder changes or when the modal toggles, not editing or actions
        return (
            this.props.folder !== nextProps.folder ||
            this.state.addStubOpen !== nextState.addStubOpen ||
            this.props.disabled !== nextProps.disabled
        );
    }

    updateDropdown() {
        this.stubContainer.style.height = (this.props.folder.open ? this.stubWrap.clientHeight : 0) + 'px';
    }

    editName() {
        this.input.disabled = false;
        this.input.focus();
        this.input.setSelectionRange(0, this.input.value.length);
        this.setState({ editing: true });
    }

    saveName(e) {
        this.input.disabled = true;
        this.props.actions.editFolder(this.props.folder.id, e.target.value);
        this.setState({ editing: false });
    }

    nameBlur(e) {
        this.saveName(e);
    }

    nameInput(e) {
        if (e.which === 13) this.input.blur();
    }

    addStub() {
        this.setState({ addStubOpen: true });
    }

    closeAddStub() {
        this.setState({ addStubOpen: false });
    }

    remove() {
        this.props.actions.removeFolder(this.props.folder);
    }

    toggle() {
        this.props.actions.toggleFolderOpen(this.props.id);
    }

    inputClick(e) {
        if (this.state.editing) e.stopPropagation();
    }

    disable() {
        this.props.actions.disableStub(this.props.folder.stubs, !this.props.disabled);
    }

    // Drag stuff, vanilla to avoid the react lifecycle
    mouseDown(e) {
        this.startY = e.clientY;
        this.lastPlacement = -1;
        this.rAF = null;

        this.offsetY = this.wrapper.getBoundingClientRect().top;
        this.parent = this.wrapper.parentNode;

        this.positions = this.findPlacements();
        console.log(this.positions);

        this.wrapper.classList.add('dragging');
        this.wrapper.style.transform = `translate3d(0, ${this.offsetY}px, 0)`;
        this.app.appendChild(this.wrapper);

        // create a placer element to indicate where the stub will be dropped
        this.placer = document.createElement('div');
        this.placer.classList.add('stub-placer');
        this.app.appendChild(this.placer);
        this.placer.style.top = this.offsetY + 'px';

        window.addEventListener('mousemove', this.mouseMove);
        window.addEventListener('mouseup', this.mouseUp);
    }

    mouseMove(e) {
        const { startY, offsetY, positions, placer, wrapper } = this;

        let diffY = e.clientY - startY + offsetY;
        const previousPlacement = this.lastPlacement;

        if (diffY > positions[0].y) {
            for (let i = 1; i < positions.length; i++) {
                let prev = positions[i - 1];
                let next = positions[i];
                if (diffY < next.y && diffY > prev.y && i !== this.lastPlacement) {
                    this.lastPlacement = i;
                    break;
                }
            }
        } else {
            this.lastPlacement = 0;
        }

        if (previousPlacement !== this.lastPlacement) {
            placer.style.top = (positions[this.lastPlacement].y) + 'px';
            console.log(positions[this.lastPlacement]);
        }

        cancelAnimationFrame(this.rAF);
        this.rAF = requestAnimationFrame(() => {
            wrapper.style.transform = `translate3d(0, ${diffY}px, 0)`;
        });
    }

    mouseUp() {
        const { lastPlacement, positions, placer, wrapper, rAF } = this;

        cancelAnimationFrame(rAF);

        window.removeEventListener('mousemove', this.mouseMove);
        window.removeEventListener('mouseup', this.mouseUp);

        wrapper.classList.remove('dragging');
        wrapper.style = '';

        placer.remove();
        this.parent.appendChild(this.wrapper);

        if (lastPlacement < 0) return;

        // check for movement
        const placement = positions[lastPlacement];
        this.props.actions.moveFolder(this.props.folder.id, placement.index);
        
        console.log('placement ', placement);
    }

    findPlacements() {
        let placements = [];
        document.querySelectorAll('.folder').forEach((folder, folderIndex) => {
            const header = folder.querySelector('.folder__header');
            const box = header.getBoundingClientRect();
            // one additional placement for before the first folder
            if (folderIndex === 0) placements.push({ y: box.top - 8, index: 0 });
            // placements are all below each folder
            placements.push({ y: box.bottom + 2 + folder.querySelector('.folder__stubs').clientHeight, index: folderIndex + 1 });
        });
        return placements;
    }
    
    render() {
        const { disabled, folder: {id, name, open, stubs} } = this.props;
        console.log(`Folder ${id} render`);
        
        const folderClass = classNames({
            'folder': true,
            'folder--open': open,
            'disabled': disabled
        });
        return (
            <div className={folderClass}>
                <div className="folder__placeholder">
                    <div className="folder__header" onClick={this.toggle} ref={elem=>this.wrapper=elem}>
                        <div className="folder__handle" onMouseDown={this.mouseDown}></div>
                        <div className="folder__header-wrap" onClick={this.inputClick}>
                            <input ref={input=>this.input=input} className="folder__name" type="text" onKeyPress={this.nameInput} onBlur={this.nameBlur} defaultValue={name} disabled />
                        </div>
                        <div className="actions" onClick={e=>e.stopPropagation()}>
                            <button className="edit" onClick={this.editName}><Tooltip text="Edit" /></button>
                            <button className="add-stub" onClick={this.addStub}><Tooltip text="Add Stub" /></button>
                            <RemoveButton onRemove={this.remove} />
                            <button className="disable" onClick={this.disable}><Tooltip text={disabled ? 'Enable' : 'Disable'} /></button>
                        </div>
                    </div>
                </div>
                <div ref={elem=>this.stubContainer=elem} className="folder__stubs">
                    <div ref={elem => this.stubWrap = elem} className="folder__stub-list">
                        {stubs.map(stubID => (
                            <Stub key={stubID} id={stubID} folderID={id} />
                        ))}
                    </div>
                </div>
                
                <Modal title="Add Stub" open={this.state.addStubOpen} onClose={this.closeAddStub}>
                    <StubForm closeModal={this.closeAddStub} folderID={id} />
                </Modal>
            </div>
        );
    }
}

export default connect(
    (state, props) => createSelector(
        [
            (state, props) => state.folders.byHash[props.id],
            (state) => state.folders.editing,
            // selector for determining if all children stubs are disabled
            (state) => {
                const folder = state.folders.byHash[props.id];
                if (folder.stubs.length === 0) return false;

                for (let i = 0; i < folder.stubs.length; i++) {
                    if (!state.stubs[folder.stubs[i]].disabled) return false;
                }
                return true;
            }
        ], (folder, editing, disabled) => ({
            folder,
            editing: editing === folder.id,
            disabled
        })
    ),
    (dispatch) => ({
        actions: bindActionCreators({ ...folderActions, addStub, disableStub }, dispatch)
    })
)(Folder);