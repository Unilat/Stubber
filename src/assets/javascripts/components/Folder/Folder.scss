.folder {
    position: relative;
    margin: 0 10px 10px 10px;
    z-index: 1;

    &--open {
        .folder__header:after {
            transform: translateY(5px) rotate(-135deg);
        }
            .folder__header-wrap {
                border-radius: 3px 3px 0 0;
                transition: border-radius 0s 0s;
            }

        .folder__stubs {
            overflow: visible;
        }

        .folder__placeholder {
            border-radius: 3px 3px 0 0;
        }
    }

    &.disabled {
        .folder__name {
            color: #666;
            text-decoration: line-through;
        }
    }
}
    .folder__placeholder {
        height: 50px;
        background-color: #eaeff3;
        border-radius: 3px;
    }
        .folder__header {
            color: #277cca;
            font-weight: bold;
            cursor: pointer;
            position: relative;
            //border-bottom: 1px solid #cedbe2;
            z-index: 1;
        
            &:after {
                content: '';
                position: absolute;
                z-index: 1;
                right: 20px;
                top: 50%;
                margin-top: -7px;
                border: 1px solid transparent;
                border-right-color: #277cca;
                border-bottom-color: #277cca;
                height: 9px;
                width: 9px;
                transform: translateY(0) rotate(45deg);
                transition: transform 0.2s ease-out;
            }

            &.dragging {
                position: fixed;
                top: 0;
                left: 10px;
                right: 10px;
                z-index: 1000;
                opacity: 0.5;
                width: auto;

                &, .folder__handle {
                    cursor: -webkit-grabbing;
                }

                .folder__header-wrap {
                    background-color: #eaeff3;
                }
            }
        }
            .folder__handle {
                position: absolute;
                width: 20px;
                height: 20px;
                top: 0;
                left: 0;
                z-index: 10;
                cursor: -webkit-grab;
                user-select: none;
            }
            .folder__header-wrap {
                padding: 5px;
                border-radius: 3px;
                overflow: hidden;
                transition: border-radius 0s 0.2s;
            }
                .folder__name {
                    font-family: 'Roboto', sans-serif;
                    color: #277cca;
                    padding: 0 10px;
                    width: auto;
                    background-color: transparent !important;
                    border-color: transparent;
                }
                .folder__name[disabled] {
                    cursor: pointer;
                }
                    .actions div {
                        padding: 0 5px;
                    }
    .folder__stubs {
        transition: height 0.2s ease-out;
        overflow: hidden;
        position: relative;
        z-index: 2;
    }

.dark {
    .folder__placeholder {
        background-color: #292f33;
    }
        .folder__header {
            color: #deefff;
            //border-bottom: 1px solid #12191d;
        
            &:after {
                border-right-color: #deefff;
                border-bottom-color: #deefff;
            }

            &.dragging .folder__header-wrap {
                background-color: #292f33;
            }
        }
        
        .folder__name {
            color: #deefff;
        }
    
    .folder.disabled .folder__name {
        color: #738494;
    }
}