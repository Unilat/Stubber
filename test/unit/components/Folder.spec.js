import React from 'react';
import { shallow } from 'enzyme';

import { Folder } from '../../../src/assets/javascripts/components/Folder/Folder.js';

describe('Folder', () => {

    const folder = {
        name: 'Test Folder',
        id: 0,
        open: true,
        stubs: []
    };
    const mockUpdateDropdown = jest.fn();
    Folder.prototype.updateDropdown = mockUpdateDropdown;
    const wrapper = shallow(<Folder key={1} id={0} folder={folder} editing={false} disabled={false} actions={{}}></Folder>);

    test('renders', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('updates the dropdown on render', () => {
        expect(mockUpdateDropdown).toHaveBeenCalledTimes(1);
    });

    test('updates the dropdown on update', () => {
        wrapper.setProps({disabled: true});
        expect(mockUpdateDropdown).toHaveBeenCalledTimes(2);
    });
});