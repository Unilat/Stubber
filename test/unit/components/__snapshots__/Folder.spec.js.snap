// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Folder renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Folder
    actions={Object {}}
    disabled={false}
    editing={false}
    folder={
      Object {
        "id": 0,
        "name": "Test Folder",
        "open": true,
        "stubs": Array [],
      }
    }
    id={0}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="folder__placeholder"
        >
          <div
            className="folder__header"
            onClick={[Function]}
          >
            <div
              className="folder__handle"
              onMouseDown={[Function]}
            />
            <div
              className="folder__header-wrap"
              onClick={[Function]}
            >
              <input
                className="folder__name"
                defaultValue="Test Folder"
                disabled={true}
                onBlur={[Function]}
                onKeyPress={[Function]}
                type="text"
              />
            </div>
            <div
              className="actions"
              onClick={[Function]}
            >
              <button
                className="edit"
                onClick={[Function]}
              >
                <Tooltip
                  text="Edit"
                />
              </button>
              <button
                className="add-stub"
                onClick={[Function]}
              >
                <Tooltip
                  text="Add Stub"
                />
              </button>
              <RemoveButton
                onRemove={[Function]}
                tooltip="Remove"
              />
              <button
                className="disable"
                onClick={[Function]}
              >
                <Tooltip
                  text="Disable"
                />
              </button>
            </div>
          </div>
        </div>,
        <div
          className="folder__stubs"
        >
          <div
            className="folder__stub-list"
          />
        </div>,
        <Modal
          onClose={[Function]}
          open={false}
          title="Add Stub"
        >
          <Connect(StubForm)
            closeModal={[Function]}
            folderID={0}
          />
        </Modal>,
      ],
      "className": "folder folder--open",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="folder__header"
            onClick={[Function]}
          >
            <div
              className="folder__handle"
              onMouseDown={[Function]}
            />
            <div
              className="folder__header-wrap"
              onClick={[Function]}
            >
              <input
                className="folder__name"
                defaultValue="Test Folder"
                disabled={true}
                onBlur={[Function]}
                onKeyPress={[Function]}
                type="text"
              />
            </div>
            <div
              className="actions"
              onClick={[Function]}
            >
              <button
                className="edit"
                onClick={[Function]}
              >
                <Tooltip
                  text="Edit"
                />
              </button>
              <button
                className="add-stub"
                onClick={[Function]}
              >
                <Tooltip
                  text="Add Stub"
                />
              </button>
              <RemoveButton
                onRemove={[Function]}
                tooltip="Remove"
              />
              <button
                className="disable"
                onClick={[Function]}
              >
                <Tooltip
                  text="Disable"
                />
              </button>
            </div>
          </div>,
          "className": "folder__placeholder",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="folder__handle"
                onMouseDown={[Function]}
              />,
              <div
                className="folder__header-wrap"
                onClick={[Function]}
              >
                <input
                  className="folder__name"
                  defaultValue="Test Folder"
                  disabled={true}
                  onBlur={[Function]}
                  onKeyPress={[Function]}
                  type="text"
                />
              </div>,
              <div
                className="actions"
                onClick={[Function]}
              >
                <button
                  className="edit"
                  onClick={[Function]}
                >
                  <Tooltip
                    text="Edit"
                  />
                </button>
                <button
                  className="add-stub"
                  onClick={[Function]}
                >
                  <Tooltip
                    text="Add Stub"
                  />
                </button>
                <RemoveButton
                  onRemove={[Function]}
                  tooltip="Remove"
                />
                <button
                  className="disable"
                  onClick={[Function]}
                >
                  <Tooltip
                    text="Disable"
                  />
                </button>
              </div>,
            ],
            "className": "folder__header",
            "onClick": [Function],
          },
          "ref": [Function],
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "folder__handle",
                "onMouseDown": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <input
                  className="folder__name"
                  defaultValue="Test Folder"
                  disabled={true}
                  onBlur={[Function]}
                  onKeyPress={[Function]}
                  type="text"
                />,
                "className": "folder__header-wrap",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "folder__name",
                  "defaultValue": "Test Folder",
                  "disabled": true,
                  "onBlur": [Function],
                  "onKeyPress": [Function],
                  "type": "text",
                },
                "ref": [Function],
                "rendered": null,
                "type": "input",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <button
                    className="edit"
                    onClick={[Function]}
                  >
                    <Tooltip
                      text="Edit"
                    />
                  </button>,
                  <button
                    className="add-stub"
                    onClick={[Function]}
                  >
                    <Tooltip
                      text="Add Stub"
                    />
                  </button>,
                  <RemoveButton
                    onRemove={[Function]}
                    tooltip="Remove"
                  />,
                  <button
                    className="disable"
                    onClick={[Function]}
                  >
                    <Tooltip
                      text="Disable"
                    />
                  </button>,
                ],
                "className": "actions",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <Tooltip
                      text="Edit"
                    />,
                    "className": "edit",
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "text": "Edit",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  "type": "button",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <Tooltip
                      text="Add Stub"
                    />,
                    "className": "add-stub",
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "text": "Add Stub",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  "type": "button",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "onRemove": [Function],
                    "tooltip": "Remove",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <Tooltip
                      text="Disable"
                    />,
                    "className": "disable",
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "text": "Disable",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  "type": "button",
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="folder__stub-list"
          />,
          "className": "folder__stubs",
        },
        "ref": [Function],
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "folder__stub-list",
          },
          "ref": [Function],
          "rendered": Array [],
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Connect(StubForm)
            closeModal={[Function]}
            folderID={0}
          />,
          "onClose": [Function],
          "open": false,
          "title": "Add Stub",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "closeModal": [Function],
            "folderID": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="folder__placeholder"
          >
            <div
              className="folder__header"
              onClick={[Function]}
            >
              <div
                className="folder__handle"
                onMouseDown={[Function]}
              />
              <div
                className="folder__header-wrap"
                onClick={[Function]}
              >
                <input
                  className="folder__name"
                  defaultValue="Test Folder"
                  disabled={true}
                  onBlur={[Function]}
                  onKeyPress={[Function]}
                  type="text"
                />
              </div>
              <div
                className="actions"
                onClick={[Function]}
              >
                <button
                  className="edit"
                  onClick={[Function]}
                >
                  <Tooltip
                    text="Edit"
                  />
                </button>
                <button
                  className="add-stub"
                  onClick={[Function]}
                >
                  <Tooltip
                    text="Add Stub"
                  />
                </button>
                <RemoveButton
                  onRemove={[Function]}
                  tooltip="Remove"
                />
                <button
                  className="disable"
                  onClick={[Function]}
                >
                  <Tooltip
                    text="Disable"
                  />
                </button>
              </div>
            </div>
          </div>,
          <div
            className="folder__stubs"
          >
            <div
              className="folder__stub-list"
            />
          </div>,
          <Modal
            onClose={[Function]}
            open={false}
            title="Add Stub"
          >
            <Connect(StubForm)
              closeModal={[Function]}
              folderID={0}
            />
          </Modal>,
        ],
        "className": "folder folder--open",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="folder__header"
              onClick={[Function]}
            >
              <div
                className="folder__handle"
                onMouseDown={[Function]}
              />
              <div
                className="folder__header-wrap"
                onClick={[Function]}
              >
                <input
                  className="folder__name"
                  defaultValue="Test Folder"
                  disabled={true}
                  onBlur={[Function]}
                  onKeyPress={[Function]}
                  type="text"
                />
              </div>
              <div
                className="actions"
                onClick={[Function]}
              >
                <button
                  className="edit"
                  onClick={[Function]}
                >
                  <Tooltip
                    text="Edit"
                  />
                </button>
                <button
                  className="add-stub"
                  onClick={[Function]}
                >
                  <Tooltip
                    text="Add Stub"
                  />
                </button>
                <RemoveButton
                  onRemove={[Function]}
                  tooltip="Remove"
                />
                <button
                  className="disable"
                  onClick={[Function]}
                >
                  <Tooltip
                    text="Disable"
                  />
                </button>
              </div>
            </div>,
            "className": "folder__placeholder",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="folder__handle"
                  onMouseDown={[Function]}
                />,
                <div
                  className="folder__header-wrap"
                  onClick={[Function]}
                >
                  <input
                    className="folder__name"
                    defaultValue="Test Folder"
                    disabled={true}
                    onBlur={[Function]}
                    onKeyPress={[Function]}
                    type="text"
                  />
                </div>,
                <div
                  className="actions"
                  onClick={[Function]}
                >
                  <button
                    className="edit"
                    onClick={[Function]}
                  >
                    <Tooltip
                      text="Edit"
                    />
                  </button>
                  <button
                    className="add-stub"
                    onClick={[Function]}
                  >
                    <Tooltip
                      text="Add Stub"
                    />
                  </button>
                  <RemoveButton
                    onRemove={[Function]}
                    tooltip="Remove"
                  />
                  <button
                    className="disable"
                    onClick={[Function]}
                  >
                    <Tooltip
                      text="Disable"
                    />
                  </button>
                </div>,
              ],
              "className": "folder__header",
              "onClick": [Function],
            },
            "ref": [Function],
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "folder__handle",
                  "onMouseDown": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <input
                    className="folder__name"
                    defaultValue="Test Folder"
                    disabled={true}
                    onBlur={[Function]}
                    onKeyPress={[Function]}
                    type="text"
                  />,
                  "className": "folder__header-wrap",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "folder__name",
                    "defaultValue": "Test Folder",
                    "disabled": true,
                    "onBlur": [Function],
                    "onKeyPress": [Function],
                    "type": "text",
                  },
                  "ref": [Function],
                  "rendered": null,
                  "type": "input",
                },
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <button
                      className="edit"
                      onClick={[Function]}
                    >
                      <Tooltip
                        text="Edit"
                      />
                    </button>,
                    <button
                      className="add-stub"
                      onClick={[Function]}
                    >
                      <Tooltip
                        text="Add Stub"
                      />
                    </button>,
                    <RemoveButton
                      onRemove={[Function]}
                      tooltip="Remove"
                    />,
                    <button
                      className="disable"
                      onClick={[Function]}
                    >
                      <Tooltip
                        text="Disable"
                      />
                    </button>,
                  ],
                  "className": "actions",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": <Tooltip
                        text="Edit"
                      />,
                      "className": "edit",
                      "onClick": [Function],
                    },
                    "ref": null,
                    "rendered": Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "class",
                      "props": Object {
                        "text": "Edit",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": [Function],
                    },
                    "type": "button",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": <Tooltip
                        text="Add Stub"
                      />,
                      "className": "add-stub",
                      "onClick": [Function],
                    },
                    "ref": null,
                    "rendered": Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "class",
                      "props": Object {
                        "text": "Add Stub",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": [Function],
                    },
                    "type": "button",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "onRemove": [Function],
                      "tooltip": "Remove",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": <Tooltip
                        text="Disable"
                      />,
                      "className": "disable",
                      "onClick": [Function],
                    },
                    "ref": null,
                    "rendered": Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "class",
                      "props": Object {
                        "text": "Disable",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": [Function],
                    },
                    "type": "button",
                  },
                ],
                "type": "div",
              },
            ],
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="folder__stub-list"
            />,
            "className": "folder__stubs",
          },
          "ref": [Function],
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [],
              "className": "folder__stub-list",
            },
            "ref": [Function],
            "rendered": Array [],
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Connect(StubForm)
              closeModal={[Function]}
              folderID={0}
            />,
            "onClose": [Function],
            "open": false,
            "title": "Add Stub",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "closeModal": [Function],
              "folderID": 0,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
